proposals/0672-declarative-xcode-projects.md

```diff
@@ -0,0 +1,92 @@
+
+The current proposal suggests to move away from manually managing `Xcodeproj` files in favor of a more declarative description. There are different tools in the OSS space that allows to describe the project declaratively and that then reads that description and generate a valid Xcodeproj accordingly.
+
+The suggested too is [Tuist.io](https://tuist.io): it allows to describe the project in Swift, leveraging type checking, syntax highlights and code completion. It also uses the default language for Apple development, which is familiar to all the iOS developers and it is executable code, providing more flexibility than a pure markup language.
+
+## Basic example
+
+You can have a look at it in action in this PR:
+
+- https://github.com/facebook/react-native/pull/37952
+
+The idea is to describe the project with a declarative approach using Swift. Tuist allows you to load the `Project.swift` file in an instance of Xcode throught the `tuist edit` command. This allows to leverage code completion, type checking and syntax highlight to ensure the correctness of the project.
+
+## Motivation
+
+Managing Xcodeproj files is error prone and time consuming. Those files are source of frustrations for teams using them and they have been reported as one of the main pain point when users have to migrate to a new version of React Native.
+
+Xcode generates projects using random UUIDs to identify the objects that constitutes a project and to refer to them in the project description file itself. These UUIDs are a nightmare to work with, especially when large teams use the same project's file: when two engineers adds two files in the same project, Xcode may create two different UUIDs for the same file and it is usually messy to handle that conflict.
+
+Managing Xcodeproj manually is also an old and not efficient way to manage projects: even Apple realized that and started developing Swift Package Manager (SwiftPM) to replace the project for developing libraries.
+
+Modern approaches rely on a declarative syntax to describe how the project should look like and they rely on tools that can generate a valid Xcodeproj starting from that description. For example, [XCodeGen](https://github.com/yonaskolb/XcodeGen) uses YAML to describe the project, while [Tuist.io](https://tuist.io) uses directly swift to achieve the same result.
+
+The proposed solution is, indeed, to use [Tuist.io](https://tuist.io). It is an OSS tool, with a vibrant community and backed and supported by several companies. Some of its maintainers are also React Native core contributors.
+
+## Detailed design
+
+To adopt Tuist in React Native, we need to follow some steps:
+
+1. Land https://github.com/facebook/react-native/pull/37952. This PR already contains a functioning project which passed all the iOS tests in CI.
+2. Update the CLI to install Tuist in the user machine. Tuist comes directly with an additional executable called `tuistenv`, which handles the environment and version for the user.
```

Comment:
Does the CLI currently *install* anything (other than `node_modules`) on the user's machine? I think we may be underestimating the complexity of automating installs across different operating systems, package managers, shells, permission levels...

A guiding principle here IMO is that if we can't automate it *reliably*, we shouldn't automate it.

Reply:
> Does the CLI currently install anything (other than node_modules) on the user's machine?
It currently installs `bundle` for Ruby.

Tuist is fairly easy to automate and also we can assume that the user is on a mac machine as you could not build iOS on other machines anyway. So this limits the configuration space sensibly! :D 

That said, @pepicrft is suggesting an alternative way (which IIUC is basically to ship the binary in a node module, avoiding to install Tuist on the user machine) because he has the same concerns. I don't really know what this path requires, but I'm more than happy to explore it!

Reply:
Could Tuist be something we recommend vs. push on users to use? Is the intention that the upgrade-helper is simpler to follow if we adopt Tuist in the template app, but the requirement being complete adoption?